src/Microsoft.DotNet.Wpf/src/Shared/MS/Internal/AppDomainShutdownMonitor.cs

```diff
@@ -30,57 +30,53 @@ static AppDomainShutdownMonitor()
         {
             AppDomain.CurrentDomain.DomainUnload += OnShutdown;
             AppDomain.CurrentDomain.ProcessExit += OnShutdown;
-
-            _dictionary = new Dictionary<WeakReference, WeakReference>();
         }
-        
-        public static void Add(WeakReference listener)
+
+        public static void Add(WeakReference<IAppDomainShutdownListener> listener)
         {
-            Debug.Assert(listener.Target != null);
-            Debug.Assert(listener.Target is IAppDomainShutdownListener);
+            Debug.Assert(listener.TryGetTarget(out _));
 
-            lock (_dictionary)
+            lock (_hashSet)
             {
                 if (!_shuttingDown)
                 {
-                    _dictionary.Add(listener, listener);
+                    _hashSet.Add(listener);
                 }
             }
         }
 
-        public static void Remove(WeakReference listener)
+        public static void Remove(WeakReference<IAppDomainShutdownListener> listener)
         {
-            Debug.Assert(listener.Target == null || listener.Target is IAppDomainShutdownListener);
-
-            lock (_dictionary)
+            lock (_hashSet)
             {
                 if (!_shuttingDown)
                 {
-                    _dictionary.Remove(listener);
+                    _hashSet.Remove(listener);
                 }
             }
         }
 
         private static void OnShutdown(object sender, EventArgs e)
         {
-            lock (_dictionary)
-            {     
+            lock (_hashSet)
+            {
                 // Setting this to true prevents Add and Remove from modifying the list. This
                 // way we call out without holding a lock (which would be bad)
                 _shuttingDown = true;
             }
-            
-            foreach (WeakReference value in _dictionary.Values)
+
+            foreach (var weakReference in _hashSet)
             {
-                IAppDomainShutdownListener listener = value.Target as IAppDomainShutdownListener;
-                if (listener != null)
+                if (weakReference.TryGetTarget(out var listener))
                 {
                     listener.NotifyShutdown();
                 }
             }
         }
-        
-        private static Dictionary<WeakReference, WeakReference> _dictionary;
+
+        private static readonly HashSet<WeakReference<IAppDomainShutdownListener>> _hashSet =
```

Comment:
this isn't (and wasn't) a very helpful name and does not convey the purpose of the field (demonstrated by the need to rename it in this case), maybe `_listeners` would be better?

Reply:
moving the initialization from constructor to initializer looks like a deliberate code style change, although I am less bothered by this one, especially as it is more correct to initialize the storage before subscribing to events